conjur_api/interface/authentication_strategy_interface.py

```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+
+"""
+AuthenticationStrategy Interface
+This class describes a shared interface for authenticating to Conjur
+"""
+
+# Builtins
+import abc
+
+from conjur_api.models.ssl.ssl_verification_metadata import SslVerificationMetadata
+
+# pylint: disable=too-few-public-methods
+class AuthenticationStrategyInterface(metaclass=abc.ABCMeta):  # pragma: no cover
+    """
+    AuthenticationStrategyInterface
+    This class is an interface that outlines a shared interface for authentication strategies
+    """
+
+    @abc.abstractmethod
+    async def authenticate(self, api_key: str, ssl_verification_data: SslVerificationMetadata) -> str:
```

Comment:
I might be missing something on the authenticate signature on this interface. Not all authenticators use the `api_key` parameter. I think they each have their own set of parameters required to carry out authentication.

Reply:
Can you give me an example of an authenticator that doesn't use an api key?

Reply:
I agree w/ @doodlesbykumbi. Only the default authenticator uses the API key.  All other authenticators validate the external identity and return a Conjur Auth token.

What about injecting the strategy along the lines of:
```py
credentials = CredentialsData(username=username, password=password, machine=conjur_url)
credentials_provider = SimpleCredentialsProvider()
credentials_provider.save(credentials)
del credentials

client = Client(connection_info,
                authentication_strategy=Authn(credentials_provider=credentials_provider)
                ssl_verification_mode=ssl_verification_mode)

client = Client(connection_info,
                authentication_strategy=AuthnLdap(credentials_provider=credentials_provider)
                ssl_verification_mode=ssl_verification_mode)

client = Client(connection_info,
                authentication_strategy=AuthnJwt(jwt='path-to-jwt')
                ssl_verification_mode=ssl_verification_mode)
```
If the Authentication Strategy interface includes a method (maybe `authenticate`), and returns a Conjur Auth Token, we can inject the "how" into our client.